File: src/handoff.js
const fs = require('node:fs');
const path = require('node:path');
const paths = require('./paths');

function findSession(profileDir, sessionId) {
  const projectRoot = path.join(profileDir, 'projects');
  if (!fs.existsSync(projectRoot)) return null;
  const queue = [projectRoot];
  while (queue.length) {
    const current = queue.shift();
    for (const entry of fs.readdirSync(current, { withFileTypes: true })) {
      const full = path.join(current, entry.name);
      if (entry.isDirectory()) queue.push(full);
      else if (entry.isFile() && entry.name === `${sessionId}.jsonl`) return full;
    }
  }
  return null;
}

function copySession(from, to, sessionId) {
  const source = findSession(paths.profileDir(from), sessionId);
  if (!source) throw new Error(`Session ${sessionId} was not found in profile '${from}'.`);
  const relative = path.relative(paths.profileDir(from), source);
  const target = path.join(paths.profileDir(to), relative);
  if (fs.existsSync(target)) throw new Error(`Target session already exists: ${target}`);
  fs.mkdirSync(path.dirname(target), { recursive: true });
  fs.copyFileSync(source, target, fs.constants.COPYFILE_EXCL);
  return { source, target };
}

module.exports = { copySession, findSession };

File: rust/src/handoff.rs
use crate::paths;
use anyhow::{bail, Result};
use std::path::PathBuf;

fn find_session(profile_dir: &std::path::Path, session_id: &str) -> Option<PathBuf> {
    let root = profile_dir.join("projects");
    if !root.exists() {
        return None;
    }
    let target_name = format!("{session_id}.jsonl");
    let mut queue = vec![root];
    while let Some(current) = queue.pop() {
        let Ok(entries) = std::fs::read_dir(&current) else { continue };
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                queue.push(path);
            } else if entry.file_name().to_string_lossy() == target_name {
                return Some(path);
            }
        }
    }
    None
}

pub fn copy_session(from: &str, to: &str, session_id: &str) -> Result<(PathBuf, PathBuf)> {
    let from_dir = paths::profile_dir(from);
    let source = find_session(&from_dir, session_id).ok_or_else(|| anyhow::anyhow!("Session {session_id} was not found in profile '{from}'."))?;
    let relative = source.strip_prefix(&from_dir)?;
    let target = paths::profile_dir(to).join(relative);
    if target.exists() {
        bail!("Target session already exists: {}", target.display());
    }
    if let Some(dir) = target.parent() {
        std::fs::create_dir_all(dir)?;
    }
    std::fs::copy(&source, &target)?;
    Ok((source, target))
}
